### PR TITLE
coverage: Clean up marker statements that aren't needed later

### PR DESCRIPTION
**Pull request by Zalathar**
Some of the marker statements used by coverage are added during MIR building for use by the InstrumentCoverage pass (during analysis), and are not needed afterwards.

@rustbot label +A-code-coverage

### CODE REVIEW

**Comment by Zalathar** [comment]
I’ve had the idea of doing this for a while, but I knew it couldn’t be done directly in InstrumentCoverage (because you can’t have nop statements that early), so I figured it would require its own separate clean pass.

But then it occurred to me to look for existing cleanup passes to piggy-back on, and this seems like a perfect fit.

**Review by Swatinem — APPROVED** [review]

**Comment by Zalathar** [comment]
Based on https://github.com/rust-lang/rust/pull/114917#issuecomment-2000686502 I suspect the `//@ ignore-windows` isn't actually needed for coverage mir-opt tests.

But I think I'll leave things as-is in this PR, and then change the ignores in a follow-up PR, so that if anything goes wrong the changes won't be mixed together.

**Comment by Zalathar** [comment]
> But I think I'll leave things as-is in this PR, and then change the ignores in a follow-up PR, so that if anything goes wrong the changes won't be mixed together.

I ended up changing this after all. The new test now runs on any target, and uses `-Zno-profiler-runtime` instead of `//@ needs-profiler-support`.

**Comment by Zalathar** [comment]
This is now a prerequisite of #122860.

It can still be reviewed independently in its own right, or it can be incorporated into the review of that PR.

File: compiler/rustc_mir_transform/src/cleanup_post_borrowck.rs [hunk 1]
```diff
@@ -25,6 +30,12 @@ impl<'tcx> MirPass<'tcx> for CleanupPostBorrowck {
                 match statement.kind {
                     StatementKind::AscribeUserType(..)
                     | StatementKind::Assign(box (_, Rvalue::Ref(_, BorrowKind::Fake, _)))
+                    | StatementKind::Coverage(box Coverage {
+                        // These kinds of coverage statements are markers inserted during
+                        // MIR building, and are not needed after InstrumentCoverage.
+                        kind: CoverageKind::BlockMarker { .. } | CoverageKind::SpanMarker { .. },
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/cleanup_post_borrowck.rs:36]: Please also add this to validation to ensure they never get reintroduced

Comment (Zalathar) [reply compiler/rustc_mir_transform/src/cleanup_post_borrowck.rs:36]: I added an appropriate check to `CfgChecker` [(diff)](https://github.com/rust-lang/rust/compare/313b6a04ed88bcbe2a47d4d9395478ddaeec986b..42d7797d02f5f0021c8904efc5f72789c834ca40).

**Comment by Zalathar** [comment]
The only change to `syntax.rs` is describing the new validation rule in the docs for `AnalysisPhase::PostCleanup`.

**Comment by Zalathar** [comment]
Oh, I guess `mir-opt` already sets `-Zvalidate-mir`, so I don't need to set it explicitly.

**Pull request merged**